pkcs11/pkcs11.go

```diff
@@ -1687,3 +1706,169 @@ func (s *Slot) generateECDSA(o keyOptions) (crypto.PrivateKey, error) {
 	}
 	return priv, nil
 }
+
+func WithHash(privKey crypto.PrivateKey, hash crypto.Hash) (crypto.PrivateKey, error) {
+	r, ok := privKey.(*rsaPrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("expected type rsaPrivateKey, but got type %T", privKey)
+	}
+	r.hash = &hash
+	return r, nil
+}
+
+func (r *rsaPrivateKey) getHash() *crypto.Hash {
+	if r.hash != nil {
+		return r.hash
+	}
+	// Initialized to SHA256 if no hash specified
+	hash := crypto.SHA256
+	return &hash
+}
+
+func WithPublicKeyHandle(privKey crypto.PrivateKey, o Object) crypto.PrivateKey {
+	r := privKey.(*rsaPrivateKey)
+	r.pubH = o.o
+	return privKey
+}
+
+func (r *rsaPrivateKey) encryptRSA(data []byte) ([]byte, error) {
+	hash := r.getHash()
+	cParam, err := makeCParamRSAOAEP(hash)
+	if err != nil {
+		return nil, err
+	}
+	defer C.free(unsafe.Pointer(cParam))
+
+	cDataBytes := toCBytes(data)
+
+	m := makeMechanism(C.CKM_RSA_PKCS_OAEP, C.CK_VOID_PTR(cParam), C.sizeof_CK_RSA_PKCS_OAEP_PARAMS)
+
+	rv := C.ck_encrypt_init(r.o.fl, r.o.h, &m, r.pubH)
+	if err := isOk("C_EncryptInit", rv); err != nil {
+		return nil, err
+	}
+
+	var cCipherLen C.CK_ULONG
+
+	// First call is used to determine maximum length of encrypted data (PKCS #11 Specifications Section 5.2)
+	rv = C.ck_encrypt(r.o.fl, r.o.h, &cDataBytes[0], C.CK_ULONG(len(cDataBytes)), nil, &cCipherLen)
+	if err := isOk("C_Encrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cCipher := make([]C.CK_BYTE, cCipherLen)
+
+	rv = C.ck_encrypt(r.o.fl, r.o.h, &cDataBytes[0], C.CK_ULONG(len(cDataBytes)), &cCipher[0], &cCipherLen)
+	if err := isOk("C_Encrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cipher := toBytes(cCipher)
+
+	return cipher, nil
+}
+
+func (r *rsaPrivateKey) decryptRSA(encryptedData []byte) ([]byte, error) {
```

Comment:
The interface this should implement is: https://pkg.go.dev/crypto#Decrypter

So something like

```
func (r *rsaPrivateKey) Decrypt(rand io.Reader, msg []byte, opts crypto.DecrypterOpts) (plaintext []byte, err error)
```

For DecrypterOpts, match the RSA private key behavior https://pkg.go.dev/crypto/rsa#PrivateKey.Decrypt

Specifically:

> If opts is nil or of type *PKCS1v15DecryptOptions then PKCS #1 v1.5 decryption is performed. Otherwise opts must have type *OAEPOptions and OAEP decryption is done.

Reply:
After making this change, how would decrypt be accessed outside of go-pkcs11? Is it still possible, considering that rsaPrivateKey is a private type?

Reply:
Yes, you can still type cast a type that is private. It is odd, but this is the same way that signing works:

https://github.com/google/go-pkcs11/blob/3bb1db427a996903b8dd7559b60857f4d17ac657/pkcs11/pkcs11.go#L1221

Reply:
Ah okay, thank you for clarifying!